### PR TITLE
persist: unbreak Kafka upsert sources

### LINKED ISSUE #9614: persist: crash with persistent kafka upsert source with a byte/text key
I'm crashing with the following log output / stack trace

```
materialized v0.12.1-dev (54b2775d7) listening on 0.0.0.0:6875...
2021-12-15T19:01:04.823846Z  INFO coord::catalog: create source materialize.public.perf_upsert_source (u1)
2021-12-15T19:01:04.824143Z  INFO coord::catalog: create index materialize.public.perf_upsert_source_primary_idx (u2)
2021-12-15T19:01:04.835249Z  INFO coord::timestamp: Timestamping Source u1 with Real Time Consistency.
2021-12-15T19:01:04.837027Z  INFO coord::timestamp: Discovered 4 new (4 total) kafka partitions for topic perf-upsert-4part3 (source u1)
2021-12-15T19:01:04.934713Z  INFO dataflow::source::kafka: Activating Kafka queue for perf-upsert-4part3 [2] (source u1/29) on worker 0
2021-12-15T19:01:04.940017Z  INFO dataflow::source::kafka: Activating Kafka queue for perf-upsert-4part3 [1] (source u1/29) on worker 3
2021-12-15T19:01:04.940073Z  INFO dataflow::source::kafka: Activating Kafka queue for perf-upsert-4part3 [3] (source u1/29) on worker 1
2021-12-15T19:01:04.940286Z  INFO dataflow::source::kafka: Activating Kafka queue for perf-upsert-4part3 [0] (source u1/29) on worker 2
2021-12-15T19:01:07.124319Z ERROR panic: index out of bounds: the len is 2 but the index is 2
thread: timely:work-2
location: src/dataflow/src/render/upsert.rs:269:25
version: 0.12.1-dev (54b2775d7ae18656805d8d92e380cf4781927c27)
backtrace:
   0: backtrace::backtrace::trace
   1: backtrace::capture::Backtrace::new
   2: materialized::handle_panic
   3: core::ops::function::Fn::call
   4: std::panicking::rust_panic_with_hook
   5: std::panicking::begin_panic_handler::{{closure}}
   6: std::sys_common::backtrace::__rust_end_short_backtrace
   7: _rust_begin_unwind
   8: core::panicking::panic_fmt
   9: core::panicking::panic_bounds_check
  10: dataflow::render::upsert::evaluate::{{closure}}
  11: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
  12: core::option::Option<T>::map
  13: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::next
  14: repr::row::Row::extend
  15: dataflow::render::upsert::evaluate
  16: dataflow::render::upsert::upsert::{{closure}}::{{closure}}
  17: core::result::Result<T,E>::and_then
  18: dataflow::render::upsert::upsert::{{closure}}
  19: <timely::dataflow::stream::Stream<S,D> as timely::dataflow::operators::map::Map<S,D>>::flat_map::{{closure}}::{{closure}}::{{closure}}::{{closure}}
  20: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
  21: core::option::Option<T>::map
  22: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::next
  23: <core::iter::adapters::fuse::Fuse<I> as core::iter::adapters::fuse::FuseImpl<I>>::next
  24: <core::iter::adapters::fuse::Fuse<I> as core::iter::traits::iterator::Iterator>::next
  25: <core::iter::adapters::flatten::FlattenCompat<I,U> as core::iter::traits::iterator::Iterator>::next
  26: <core::iter::adapters::flatten::FlatMap<I,U,F> as core::iter::traits::iterator::Iterator>::next
  27: timely::dataflow::channels::pushers::buffer::Session<T,D,P>::give_iterator
  28: <timely::dataflow::stream::Stream<S,D> as timely::dataflow::operators::map::Map<S,D>>::flat_map::{{closure}}::{{closure}}::{{closure}}
  29: timely::dataflow::operators::generic::handles::InputHandle<T,D,P>::for_each
  30: <timely::dataflow::stream::Stream<S,D> as timely::dataflow::operators::map::Map<S,D>>::flat_map::{{closure}}::{{closure}}
  31: <timely::dataflow::stream::Stream<G,D1> as timely::dataflow::operators::generic::operator::Operator<G,D1>>::unary::{{closure}}::{{closure}}
  32: timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build::{{closure}}::{{closure}}
  33: timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build_reschedule::{{closure}}
  34: <timely::dataflow::operators::generic::builder_raw::OperatorCore<T,L> as timely::scheduling::Schedule>::schedule
  35: timely::progress::subgraph::PerOperatorState<T>::schedule
  36: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
  37: <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
  38: timely::progress::subgraph::PerOperatorState<T>::schedule
  39: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
  40: <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
  41: timely::worker::Wrapper::step::{{closure}}
  42: core::option::Option<T>::map
  43: timely::worker::Wrapper::step
  44: timely::worker::Worker<A>::step_or_park
  45: dataflow::server::Worker<A>::run
  46: dataflow::server::serve::{{closure}}
  47: timely::execute::execute::{{closure}}
  48: timely_communication::initialize::initialize_from::{{closure}}
  49: std::sys_common::backtrace::__rust_begin_short_backtrace
  50: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
  51: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  52: std::panicking::try::do_call
  53: ___rust_try
  54: std::panicking::try
  55: std::panic::catch_unwind
  56: std::thread::Builder::spawn_unchecked::{{closure}}
  57: core::ops::function::FnOnce::call_once{{vtable.shim}}
  58: std::sys::unix::thread::Thread::new::thread_start
  59: __pthread_start
  ```
  
  in debug immediately after ingesting the first few records. This runs fine without persisted sources. I'm on main (54b2775d7ae18656805d8d92e380cf4781927c27)
  
  + the follwing diff
  
  ```
  diff --git a/src/persist/src/operators/stream.rs b/src/persist/src/operators/stream.rs
index 9435735a9..0b2641a39 100644
--- a/src/persist/src/operators/stream.rs
+++ b/src/persist/src/operators/stream.rs
@@ -454,6 +454,9 @@ where
                                         pending_future.time,
                                     );

+                                    if pending_future.time > 5 {
+                                        let _ = condition_write.allow_compaction(Antichain::from_elem(pending_future.time - 5));
+                                    }
                                     let future = primary_write.seal(pending_future.time);

                                     pending_futures.push_back(PendingSealFuture::PrimarySeal(
@@ -523,6 +526,9 @@ where
                         PendingSealFuture::PrimarySeal(mut pending_future) => {
                             match Pin::new(&mut pending_future.future).poll(&mut context) {
                                 std::task::Poll::Ready(Ok(_)) => {
+                                    if pending_future.time > 5 {
+                                        let _ = primary_write.allow_compaction(Antichain::from_elem(pending_future.time - 5));
+                                    }
                                     log::trace!(
                                         "In {}, finished sealing primary collection up to {}",
                                         &operator_name,
diff --git a/test/performance/perf-upsert/src/main.rs b/test/performance/perf-upsert/src/main.rs
index d5ad80600..9c45380b7 100644
--- a/test/performance/perf-upsert/src/main.rs
+++ b/test/performance/perf-upsert/src/main.rs
@@ -151,7 +151,7 @@ async fn create_upsert_text_source(
             query_prefix
         )
     } else {
-        format!("{} FORMAT BYTES ENVELOPE NONE", query_prefix)
+        format!("{} FORMAT BYTES ENVELOPE UPSERT", query_prefix)
     };
     log::debug!("creating source=> {}", query);
```

I'm running materialize with

```
cargo run -- --dev -w 4 --log-file=stderr --experimental --persistent-kafka-upsert-source
```

and testing it with the perf-upsert test (test/performance/perf-upsert)

run with 

```
RUST_LOG=debug cargo run --release -- --message-count 400 --messages-per-second 50 --create-topic --kafka-topic=perf-upsert-4part3 --partitions 4 --replication-factor 1
```

### PR DESCRIPTION
**Pull request by ruchirK**
Previously, we added kafka metadata into the upsert dataflow. Unfortunately, we only did
so for the non-persisted kafka upsert dataflow, and not for the persisted kafka upsert
dataflow.

This PR aims to fix that by simply folding the metadata into the value which is what
downstream `evaluate` seems to want anyway. The existing upsert implementation likely
does not do so as a performance optimization (no need to create a bunch of new, wider
values if we suspect many of them will be discarded).

This fix is admittedly brittle. A longer term fix would be to unify more of the
code around the persisted codepath, and the nonpersisted codepath. We are leaving
that for followup work.

Fixes #9614 

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Comment by ruchirK** [comment]
I'm not at all wedded to this fix -- this is the fastest fix I could come up with to unblock some testing for feature work. Happy to scrap this and do something different if we would rather do that.

**Comment by ruchirK** [comment]
Reposting from a conversation in slack with @aljoscha (with some amendments):

I know this is a sort of hacky fix, and one potential better fix is to reduce the code duplication around `upsert_core` and `persisted_upsert` <somehow> (I think there's multiple approaches to doing so). I tried that yesterday, and quickly realized its more complicated than I had hoped. I pushed out this PR to have _some_ bandaid fix that we could be confident about because I wasn't sure how long a more proper fix would take / how long it would take to get through code reviews and I wanted to have something we could push out before everyone leaves for the holidays.

If folks feel like the actual fix is not actually that hairy, I'm more than happy to close out this PR :)

**Comment by aljoscha** [comment]
Ah, I now remember why it's hard. We require a `Codec` impl for the input/output types of `persistent_upsert()`, while we don't have (or want) `Codec` impls for the types that flow through the old `upsert_core()`. The crux is that `upsert_core()` returns a `Result<_, DataflowError>` and we don't want to get into the business of having a `Codec` for `DataflowError`.

I think we should go forward with the easy fix and also expand our test coverage to also regularly test the second (persistent) upsert implementation.

**Comment by ruchirK** [comment]
Added some test coverage to the existing CI tests -- @philip-stoev I'll defer to you if you want to make any changes I just copied a relevant subset of the existing kafka-upsert-sources.td

File: test/persistence/kafka-sources/upsert-modification-before.td
Comment (quodlibetor) [inline test/persistence/kafka-sources/upsert-modification-before.td:55]: Could you modify this to look like:

```suggestion
> CREATE MATERIALIZED SOURCE texttext
  FROM KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'testdrive-textbytes-${testdrive.seed}'
   FORMAT TEXT
   INCLUDE PARTITION AS kafka_partition, OFFSET AS mz_offset
   ENVELOPE UPSERT
```

**Review by philip-stoev — APPROVED** [review]
Thank you very much for the tests! They look perfect to me.

File: test/persistence/kafka-sources/upsert-modification-before.td
Comment (ruchirK) [reply test/persistence/kafka-sources/upsert-modification-before.td]: done!

**Comment by ruchirK** [comment]
This is RFAL!

**Comment by ruchirK** [comment]
@aljoscha and @quodlibetor could you please look at this today? I'd love to get this fix merged soon if possible. Thanks!

**Review by aljoscha — APPROVED** [review]
LGTM! 🙌

**Review by petrosagg — APPROVED** [review]

**Pull request merged**
